test/ServiceManagerTest.php

```
@@ -283,4 +253,92 @@ public function testFactoryMayBeStaticMethodDescribedByCallableString()
+
+        // will be true if factory array is properly setup
+        $this->assertTrue($sm->has('factory'));
+        $this->assertInstanceOf(InvokableObject::class, $sm->get('factory'));
+
+        // will be true if initializer is present
+        $this->assertTrue($sm->get('factory')->initializerPresent);
+
+        // will succeed if invokable is properly set up
+        $this->assertTrue($sm->has('invokable'));
+        $this->assertInstanceOf(stdClass::class, $sm->get('invokable'));
+
+        // will be true if initializer is present
+        $this->assertTrue($sm->get('invokable')->initializerPresent);
+
+        // will succeed if abstract factory is available
+        $this->assertTrue($sm->has('foo'));
+        $this->assertInstanceOf(Foo::class, $sm->get('foo'));
+
+        // will be true if initializer is present
+        $this->assertTrue($sm->get('foo')->initializerPresent);
+    }
+
+    public function testInvokablesShouldNotOverrideFactoriesAndDelegators()
+    {
+        $sm = new ServiceManager([
+            'factories' => [
+                // produce InvokableObject
+                'factory1' => SampleFactory::class,
+                'factory2' => SampleFactory::class,
+            ],
+            'delegators' => [
+                'factory1' => [
+                    // produce tagged invokable object
+                    TaggingDelegatorFactory::class,
+                ]
+            ]
+        ]);
+
+        $object1 = $sm->build('factory1');
+        // assert delegated object is produced by delegator factory
+        $this->assertTrue(isset($object1->delegatorTag));
+        $this->assertInstanceOf(InvokableObject::class, $object1);
+
+
+        $object2 = $sm->build('factory2');
+        // assert delegated object is produced by SampleFactory
+        $this->assertFalse(isset($object2->delegatorTag));
+        $this->assertInstanceOf(InvokableObject::class, $object2);
+
+        $sm->setInvokableClass('factory1', stdClass::class);
```

This should raise an exception due to already set service